### PR TITLE
Tune the AI performance at different difficulty levels

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
Related to #8270

1. On certain difficulties, allow AI to develop castles only on even days (i.e. every two days), what should slow down its development. Such difficulty is currently the Easy difficulty (for non-campaign games);
2. On certain difficulties, prohibit AI from building level 6 dwellings (but allow AI to upgrade them if they are already built). Currently, AI cannot build these dwellings on Easy difficulty (for non-campaign games);
3. ~On certain difficulties, prohibit AI from exchanging armies and artifacts between heroes. Currently it is only disabled on Easy difficulty~ similar logic is already present, but in a different form, so I removed this one;
4. Added infrastructure to support negative per-dwelling growth bonuses. Not used yet due to https://github.com/ihhub/fheroes2/discussions/8270#discussioncomment-8147181, both on Easy and Normal difficulties growth bonus is zero (partly because of point 2), but this can be tuned if needed.

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
Hi @ihhub @Branikolog @zenseii @Districh-ru there are my proposals regarding #8270. What do you think?

**Comment by oleg-derevenetz** [comment]
@zenseii well, that's need to be tested of course. At the very least, we can introduce special rules for the campaign-type games.

**Comment by oleg-derevenetz** [comment]
Hi @zenseii after a few tests, I decided to implement separate rules for campaign and non-campaign games, please see updated version of the PR description.

**Comment by Districh-ru** [comment]
> Hi @ihhub @Branikolog @zenseii @Districh-ru there are my proposals regarding #8270. What do you think?

Hi @oleg-derevenetz, I agree with the proposed changes. This will slow down enemy AI development on Easy and Normal difficulties giving new players some more time.
I'm only concerned on prohibiting AI from building level 6 dwellings on Normal difficulty. Maybe we should instead limit (if we can) the maximum number of monsters recruited from this dwelling per week... or maximum number per hero.

Also should we tweak the changes made in https://github.com/ihhub/fheroes2/pull/8256 to reduce the enemy AI aggressiveness against player on Easy and possibly Normal difficulties?

**Comment by oleg-derevenetz** [comment]
Hi @Districh-ru

> Maybe we should instead limit (if we can) the maximum number of monsters recruited from this dwelling per week... or maximum number per hero.

The problem with such approaches is that then a human player, with some luck (if there is no plague), will capture a castle with a large number of non-hired level 6 creatures. In general, I'm thinking about leaving a limit on creatures of the maximum level for Easy difficulty (and maybe remove the restriction for lvl 5 creatures for this difficulty), because level 6 creatures in HoMM2 are way too strong in comparison with level 5.

**Comment by Districh-ru** [comment]
> The problem with such approaches is that then a human player, with some luck (if there is no plague), will capture a castle with a large number of non-hired level 6 creatures. In general, I'm thinking about leaving a limit on creatures of the maximum level for Easy difficulty only (and maybe remove the restriction for lvl 5 creatures for this difficulty), because level 6 creatures in HoMM2 are way too strong in comparison with level 5.

And can we limit the growth per week for these creatures on Easy (and probably Normal) difficulties?

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz 

No 6 lvl creatures for "easy mode" seems reasonable for me. But playing for months without opposing lvl 6 creature could be boring, though. Maybe reducing their growth (for easy) is a way to go. I need a deeper testing of it...
For "normal" I prefer AI to use all game content i.e. all 6 creature tiers. 👍 Not sure we shoud cut the growth for "Normal", though.

To make it a bit harder for AI to build 6 tier on "Normal" we can forbid him using marketplaces for resource trading, how do you think? (Not sure, how it works currently.)

**Comment by oleg-derevenetz** [comment]
> No 6 lvl creatures for "easy mode" seems reasonable for me. But playing for months without opposing lvl 6 creature could be boring, though.

I thought that you wanted to have the difficulty level "for children", etc, and level 6 creatures are extremely tough in HoMM2.

> To make it a bit harder for AI to build 6 tier on "Normal" we can forbid him using marketplaces for resource trading, how do you think? (Not sure, how it works currently.)

I would not like to disable the marketplace usage even on Easy difficulty because there are resource-scarce maps where AI would be difficult without marketplace. Instead, I added the mechanism to slow down the AI castle development by limiting period between building of castle buildings - it can be "once in 2 days" like now on Easy difficulty, "once in 3 days", etc. Concerning the Normal difficulty, there are opinions that we should make it full-featured, but without any benefits for both human players and AI, which I support in principle.

**Comment by Branikolog** [comment]
@oleg-derevenetz 

>Concerning the Normal difficulty, there are opinions that we should make it full-featured, but without any benefits for both human players and AI, which I support in principle.

So am I.

>I thought that you wanted to have the difficulty level "for children", etc, and level 6 creatures are extremely tough in HoMM2.

Indeed. I need to make a couple tests more, if there's too much child-friendly behaviour without 6 tier. :)

>Instead, I added the mechanism to slow down the AI castle development by limiting period between building of castle buildings - it can be "once in 2 days" like now on Easy difficulty, "once in 3 days", etc.

Yeah, I like that proposal you've initially implemented in this PR. I believe this would help making "Easy" as more friendly mode.

> I would not like to disable the marketplace usage even on Easy difficulty because there are resource-scarce maps where AI would be difficult without marketplace. 

Correct me, if I'm wrong, but some time ago AI couldn't use markets at all and still perform quite good. :)

**Comment by oleg-derevenetz** [comment]
> Correct me, if I'm wrong, but some time ago AI couldn't use markets at all and still perform quite good. :)

I remember complaints about some maps, e.g. "Marooned" or something like this when too strong wandering monsters were rolled and AI just stuck without necessary resources, so I decided to implement the marketplace support. I would not like to disable essential game mechanisms of this kind, since we have it. It's one thing to slow down the development of AI and another to make a complete fool out of it if it was unlucky and couldn't take the right mine.

**Comment by Branikolog** [comment]
@oleg-derevenetz 

No problem at all. That was a nice improvement and I like it. 👍 

Maybe just for "Easy" AI can "forget" about markets and rely only on resources he picked up or gained from mines. After all, maps were designed by a creators to provide a +/- reasonable amount of resources for each player (remember, AI players were quite dumb in the OG), so I believe there could be found hardly any map, where AI was left withour resources and stuck completely on a starting location. Also, AI would save a ton of resources because he won't build 6 tier creatures. ;)

So it's up to you not to cut this mechanic. 👌

**Comment by oleg-derevenetz** [comment]
@Mr-Bajs

> If the marketplace is to powerfull, why not just tax it rather than disable it, similar to the unit growth and resources? Making the conversion rate for the AI less efficient?

I would not tell that a single marketplace at the beginning is very powerful :) Its prices are such that fines are not needed there, and if AI lacks resources and is forced to use the market, then its development will be slower in any case than if it has the necessary resources. Let's not rush into imposing fines for everything that comes to hand.

**Review by Branikolog — APPROVED** [review]
Nice step on making the game more flexible. I hope we can go into tuning difficulty even further.

File: src/fheroes2/ai/normal/ai_normal_castle.cpp [hunk 1]
```diff
@@ -138,6 +140,10 @@ namespace AI
     bool Build( Castle & castle, const std::vector<BuildOrder> & buildOrderList, const uint32_t multiplier = 1 )
     {
         for ( const BuildOrder & order : buildOrderList ) {
+            if ( !Difficulty::allowAIToBuildCastleBuilding( Game::getDifficulty(), Game::isCampaign(), order.building ) ) {
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/ai/normal/ai_normal_castle.cpp]: Sure, why not.

**Review by ihhub — APPROVED** [review]

**Review by Districh-ru — APPROVED** [review]
I like the changes, they will help novice players to become familiar with the game.

**Comment by ihhub** [comment]
@oleg-derevenetz , thanks a lot for these adjustments!

**Pull request merged**